# What type of apps most profitable

This project is part of the course of Data Analyst course in Dataquest. 
The Biz Model is we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. The purpose of this project is to find what types of apps is most profitable on the 2 platforms of App Store and Google Play. This would be based on the number of downloads for the 2 platforms. The data source can be obtained in this link 
- For App Store https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps
- For Google Play Store https://www.kaggle.com/lava18/google-play-store-apps
At this point of the lesson, we are not yet using numpy nor pandas and therefore the solution is a big longer. 

# Opening and Exploring the Data

Let's open and explore the data in the 2 datasets

In [1]:
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)

apple_header = apps_data[0] #creating list apple_header made up of the first row
apple = apps_data[1:] #creating a list made up of second row onwards

In [2]:
apple[0:3]

[['1',
  '281656475',
  'PAC-MAN Premium',
  '100788224',
  'USD',
  '3.99',
  '21292',
  '26',
  '4',
  '4.5',
  '6.3.5',
  '4+',
  'Games',
  '38',
  '5',
  '10',
  '1'],
 ['2',
  '281796108',
  'Evernote - stay organized',
  '158578688',
  'USD',
  '0',
  '161065',
  '26',
  '4',
  '3.5',
  '8.2.2',
  '4+',
  'Productivity',
  '37',
  '5',
  '23',
  '1'],
 ['3',
  '281940292',
  'WeatherBug - Local Weather, Radar, Maps, Alerts',
  '100524032',
  'USD',
  '0',
  '188583',
  '2822',
  '3.5',
  '4.5',
  '5.0.0',
  '4+',
  'Weather',
  '37',
  '5',
  '3',
  '1']]

In [3]:
opened_file = open('googleplaystore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)

android_header = apps_data[0] #creating list android_header made up of the first row
android = apps_data[1:] #creating a list made up of second row onwards

In [4]:
android[0:3]

[['Photo Editor & Candy Camera & Grid & ScrapBook',
  'ART_AND_DESIGN',
  '4.1',
  '159',
  '19M',
  '10,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'January 7, 2018',
  '1.0.0',
  '4.0.3 and up'],
 ['Coloring book moana',
  'ART_AND_DESIGN',
  '3.9',
  '967',
  '14M',
  '500,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design;Pretend Play',
  'January 15, 2018',
  '2.0.0',
  '4.0.3 and up'],
 ['U Launcher Lite – FREE Live Cool Themes, Hide Apps',
  'ART_AND_DESIGN',
  '4.7',
  '87510',
  '8.7M',
  '5,000,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'August 1, 2018',
  '1.2.4',
  '4.0.3 and up']]

Let's print the first few rows and find out how many rows and columns there are.

In [5]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [6]:
explore_data(apple, 0, 3, True)

['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


Number of rows: 7197
Number of columns: 17


In [7]:
explore_data(android, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


This is the header of the Apple dataset

In [8]:
apple_header

['',
 'id',
 'track_name',
 'size_bytes',
 'currency',
 'price',
 'rating_count_tot',
 'rating_count_ver',
 'user_rating',
 'user_rating_ver',
 'ver',
 'cont_rating',
 'prime_genre',
 'sup_devices.num',
 'ipadSc_urls.num',
 'lang.num',
 'vpp_lic']

And this is the header of Android dataset

In [9]:
android_header

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

# Cleaning Data by Deleting Wrong Data

There is data duplicity within the dataset from the forum in https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015. "The entry is actually missing its "Category" value, which causes the column shift"
To find out if there is wrong data, we compare the length of each rows versus the header 

In [10]:
for row in android:
    headerlength = len(android_header)
    rowlength = len(row) 
    if rowlength != headerlength:
        print(row)
        print(android.index(row))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10472


From here, we can see that the duplicate Row is in Row 10472. We will delete data in Row 10472

In [11]:
del android[10472]

Now, we will just check to make sure that it is deleted

In [12]:
explore_data(android, 10471, 10473, rows_and_columns=False)

['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']




We do the same check for Apple App store https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion. There does not appear to have any wrong data in Apple Dataset

In [13]:
for row in apple:
    headerlength = len(apple_header)
    rowlength = len(row) 
    if rowlength != headerlength:
        print(row)
        print(android.index(row))

# Cleaning Data by Removing Duplicate Entries

One potential data issue is data duplicity. Let's check if there are duplicate entries here. 

In [14]:
for app in android:
    name = app[0]
    if name == 'Twitter':
        print(app)

['Twitter', 'NEWS_AND_MAGAZINES', '4.3', '11667403', 'Varies with device', '500,000,000+', 'Free', '0', 'Mature 17+', 'News & Magazines', 'August 6, 2018', 'Varies with device', 'Varies with device']
['Twitter', 'NEWS_AND_MAGAZINES', '4.3', '11667403', 'Varies with device', '500,000,000+', 'Free', '0', 'Mature 17+', 'News & Magazines', 'August 6, 2018', 'Varies with device', 'Varies with device']
['Twitter', 'NEWS_AND_MAGAZINES', '4.3', '11657972', 'Varies with device', '500,000,000+', 'Free', '0', 'Mature 17+', 'News & Magazines', 'July 30, 2018', 'Varies with device', 'Varies with device']


From the above is evident there are duplicity. We will remove the duplicate entries to keep only the unique entries. To remove the duplicates, we need a criteria for removing. From the above, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show the data was collected at different times. We can use this information to build a criterion for removing the duplicates. The higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, we'll only keep the row with the highest number of reviews and remove the other entries for any given app.

We will start by creating a dictionary where each key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app.

In [15]:
android_unique = []
android_duplicate = []

for app in android: #we loop through android dataset
    app_name = app[0]
    if app_name in android_unique: #if the name is in android_unique dataset we append name of the app to android_duplicate
        android_duplicate.append(app_name)
    else:
        android_unique.append(app_name)

print("android unique:",len(android_unique))
print("android duplicate:",len(android_duplicate))

print(android_unique[0:3]) #to do a check on the list of unique android apps
print('\n')
print(android_duplicate[0:3]) #to do a check on the list of duplicate android apps

android unique: 9659
android duplicate: 1181
['Photo Editor & Candy Camera & Grid & ScrapBook', 'Coloring book moana', 'U Launcher Lite – FREE Live Cool Themes, Hide Apps']


['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business']


In [16]:
print("Expected length:", len(android) - 1181)

Expected length: 9659


We will create a dictionary where each key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app. After removing the duplicates, we should be left with 9659 rows. 

To remove the duplicates, we will:
- Create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app.
- Use the information stored in the dictionary and create a new data set, which will have only one entry per app (and for each app, we'll only select the entry with the highest number of reviews). The assumption is that the highest number of reviews would be the most recent and therefore most accurate. 

---------------------------------------------------------------------------------------------------------------------

- We will start by creating an empty dictionary named reviews_max.
- Loop through the Google Play data set. For each iteration:
  - Assign the app name to a variable named name.
  - Convert the number of reviews to float. Assign it to a variable named n_reviews.
  - If name already exists as a key in the reviews_max dictionary and reviews_max[name] < n_reviews, update the  
    number of reviews for that entry in the reviews_max dictionary.
- If name is not in the reviews_max dictionary as a key, create a new entry in the dictionary where the key is the app name, and the value is the number of reviews. 

In [17]:
reviews_max = {}

for app in android:
    name = app[0] 
    n_reviews = float(app[3])
    #if the name is in the dictionary and no. of reviews is smaller than the reviews in the dataset
    if name in reviews_max and reviews_max[name] < n_reviews: 
        reviews_max[name] = n_reviews
    #if the name is not in the reviews, we will insert in the key and the value
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

print("android",len(reviews_max)) #to check number of unique apps

android 9659


We will use the dictionary created above to remove the duplicate rows

- Start by creating two empty lists: android_clean (which will store our new cleaned data set) and already_added (which will just store app names).
- Loop through the Google Play data set, and for each iteration:
  - Assign the app name to a variable named name.
  - Convert the number of reviews to float, and assign it to a variable named n_reviews.
- If n_reviews is the same as the number of maximum reviews of the app name (the number can be found in the reviews_max dictionary) and name is not already in the list already_added (read the solution notebook to find out why we need this supplementary condition):
  - Append the entire row to the android_clean list (which will eventually be a list of list and store our 
    cleaned data set).
  - Append the name of the app name to the already_added list — this helps us to keep track of apps that we 
    already added.

android_clean list should have 9659 rows.

In [18]:
#store our new cleaned data set
android_clean = []
#which will just store app names
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        #a list which will eventually be a list of list and store our cleaned data set
        android_clean.append(app)
        #Append the name of the app name to the already_added list to keep track of apps that we already added.
        already_added.append(name)

print(len(android_clean))

9659


We do the same for iOS apps as well.

In [19]:
apple_unique = []
apple_duplicate = []

for app in apple: 
    app_name = app[2]
    if app_name in apple_unique: #if the name is in apple_unique dataset we append name of the app to apple_duplicate
        apple_duplicate.append(app_name)
    else:
        apple_unique.append(app_name)

print("apple unique:",len(apple_unique))
print("apple duplicate:",len(apple_duplicate))

print(apple_unique[0:3])
print('\n')
print(apple_duplicate[0:3])

apple unique: 7195
apple duplicate: 2
['PAC-MAN Premium', 'Evernote - stay organized', 'WeatherBug - Local Weather, Radar, Maps, Alerts']


['VR Roller Coaster', 'Mannequin Challenge']


In [20]:
print("Expected length:", len(apple) - 2)

Expected length: 7195


In [21]:
reviews_max = {}

for app in apple:
    name = app[2] 
    n_reviews = float(app[6])
    #if the name is in the dictionary and no. of reviews is smaller than the reviews in the dataset
    if name in reviews_max and reviews_max[name] < n_reviews: 
        reviews_max[name] = n_reviews
    #if the name is not in the reviews, we will insert in the key and the value
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

print("apple",len(reviews_max)) #to check number of unique apps

apple 7195


In [22]:
#store our new cleaned data set
apple_clean = []
#which will just store app names
already_added = []

for app in apple:
    name = app[2]
    n_reviews = float(app[6])
    if n_reviews == reviews_max[name] and name not in already_added:
        #a list which will eventually be a list of list and store our cleaned data set
        apple_clean.append(app)
        #Append the name of the app name to the already_added list to keep track of apps that we already added.
        already_added.append(name)

print(len(apple_clean))

7195


# Removing Non-English Apps

Studying the data, we find that are some non-English apps. Example as below

In [23]:
print(android_clean[4412][0])
print(android_clean[7940][0])

中国語 AQリスニング
لعبة تقدر تربح DZ


We'd like to analyze only the apps that are directed toward an English-speaking audience. One way to go about this is to remove each app with a name containing a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /). The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII (American Standard Code for Information Interchange) system. Based on this number range, we can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.

In [24]:
def Engcheck(strinput):
    for iter in strinput:
        print(iter)
    if (ord(iter)>127):
        return False
    else:
        return True

Let's check if it works.

In [25]:
Engcheck('Instagram')

I
n
s
t
a
g
r
a
m


True

In [26]:
Engcheck('爱奇艺PPS -《欢乐颂2》电视剧热播')

爱
奇
艺
P
P
S
 
-
《
欢
乐
颂
2
》
电
视
剧
热
播


False

In [27]:
Engcheck('Docs To Go™ Free Office Suite')

D
o
c
s
 
T
o
 
G
o
™
 
F
r
e
e
 
O
f
f
i
c
e
 
S
u
i
t
e


True

In [28]:
Engcheck('Instachat 😜')

I
n
s
t
a
c
h
a
t
 
😜


False

It passes for identifying non-English letters, but emoji is identifed as non-English. To address that, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This means all English apps with up to three emoji or other special characters will still be labeled as English

In [29]:
def Engcheck3x(strinput):
    counter = 0
    for iter in strinput:
        if (ord(iter)>127):
            counter += 1
    if counter>3:
        return False
    else:
        return True

Let's test it out again with Instachat 😜 which failed previously. 

In [30]:
Engcheck3x('Instachat 😜')

True

Let's use that to filter out the non-English from our earlier list. 

In [31]:
android_eng = []
android_xeng = []

for app in android_clean:
    name = app[0]
    
    if Engcheck3x(name) == True:
        android_eng.append(app)
    else:
        android_xeng.append(app)

print(len(android_eng))
print(len(android_xeng))

9614
45


In [32]:
apple_eng = []
apple_xeng = []

for app in apple_clean:
    name = app[2]    
    if Engcheck3x(name) == True:
        apple_eng.append(app)
    else:
        apple_xeng.append(app)

print(len(apple_eng))
print(len(apple_xeng))

6181
1014


# Isolating the Free Apps

We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps; we'll need to isolate only the free apps for our analysis. We'll loop through each data set to isolate the free apps in separate lists.

In [33]:
android_final = []

for app in android_eng:
    pricing = app[6]
    if pricing == 'Free':
        android_final.append(app)

print("android final clean list:",len(android_final))

android final clean list: 8863


The treatment of identification of Free is different for iOS apps, and we will use the pricing is zero to identify. 

In [34]:
apple_final = []

for app in apple_eng:
    pricing = app[5]
    if pricing == '0':
        apple_final.append(app)

print("apple final clean list:",len(apple_final))

apple final clean list: 3220


# Most Common Apps by Genre

Our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps. Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. Let's begin the analysis by getting a sense of what are the most common genres for each market. For this, we'll need to build frequency tables for a few columns in our data sets.

In [35]:
android_header

['App',
 'Category',
 'Rating',
 'Reviews',
 'Size',
 'Installs',
 'Type',
 'Price',
 'Content Rating',
 'Genres',
 'Last Updated',
 'Current Ver',
 'Android Ver']

In [36]:
android[8:10]

[['Garden Coloring Book',
  'ART_AND_DESIGN',
  '4.4',
  '13791',
  '33M',
  '1,000,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'September 20, 2017',
  '2.9.2',
  '3.0 and up'],
 ['Kids Paint Free - Drawing Fun',
  'ART_AND_DESIGN',
  '4.7',
  '121',
  '3.1M',
  '10,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design;Creativity',
  'July 3, 2018',
  '2.8',
  '4.0.3 and up']]

In [37]:
apple_header

['',
 'id',
 'track_name',
 'size_bytes',
 'currency',
 'price',
 'rating_count_tot',
 'rating_count_ver',
 'user_rating',
 'user_rating_ver',
 'ver',
 'cont_rating',
 'prime_genre',
 'sup_devices.num',
 'ipadSc_urls.num',
 'lang.num',
 'vpp_lic']

In [38]:
apple[8:10]

[['9',
  '284666222',
  'PCalc - The Best Calculator',
  '49250304',
  'USD',
  '9.99',
  '1117',
  '4',
  '4.5',
  '5',
  '3.6.6',
  '4+',
  'Utilities',
  '37',
  '5',
  '1',
  '1'],
 ['10',
  '284736660',
  'Ms. PAC-MAN',
  '70023168',
  'USD',
  '3.99',
  '7885',
  '40',
  '4',
  '4',
  '4.0.4',
  '4+',
  'Games',
  '38',
  '0',
  '10',
  '1']]

From the above, we'll use the column Genres (column no. 9) for Android, and Prime Genre (column no. 12) for Apple. 

We'll build two functions we can use to analyze the frequency tables:
- One function to generate frequency tables that show percentages
- Another function we can use to display the percentages in a descending order. 

Dictionaries don't have order, and it will be very difficult to analyze the frequency tables. We'll need to build a second function which can help us display the entries in the frequency table in a descending order. To do that, we'll need to make use of the built-in sorted() function. This function takes in an iterable data type (like a list, dictionary, tuple, etc.), and returns a list of the elements of that iterable sorted in ascending or descending order (the reverse parameter controls whether the order is ascending or descending).

The sorted() function doesn't work too well with dictionaries because it only considers and returns the dictionary keys. However, the sorted() function works well if we transform the dictionary into a list of tuples, where each tuple contains a dictionary key along with its corresponding dictionary value. To ensure the sorting works right, the dictionary value comes first, and the dictionary key comes second. 

To start, we are going to create a Frequency Table function. The Frequency Table takes in two inputs: dataset (which is expected to be a list of lists) and index (which is expected to be an integer). The function should return the frequency table (as a dictionary) for any column we want. The frequencies should also be expressed as percentages.

In [39]:
#creating the frequency table
def freq_table(dataset,index):
    listoflist = {}
    total = 0
    for counter in dataset:
        total +=1
        value = counter[index]
        if value in listoflist:          
            listoflist[value] += 1
        else:
            listoflist[value] = 1 
    
    #showing as percentages
    listoflist_percentages = {}
    for key in listoflist:
        percentage = (listoflist[key] / total) * 100
        listoflist_percentages[key] = percentage 
    
    return listoflist_percentages

In [40]:
#displaying the table in descending order
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key) #the dictionary value comes first, and the dictionary key comes second.
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True) #sorting by value
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [41]:
display_table(apple_final,12)

Games : 58.13664596273293
Entertainment : 7.888198757763975
Photo & Video : 4.968944099378882
Education : 3.6645962732919255
Social Networking : 3.291925465838509
Shopping : 2.608695652173913
Utilities : 2.515527950310559
Sports : 2.142857142857143
Music : 2.049689440993789
Health & Fitness : 2.018633540372671
Productivity : 1.7391304347826086
Lifestyle : 1.5838509316770186
News : 1.3354037267080745
Travel : 1.2422360248447204
Finance : 1.1180124223602486
Weather : 0.8695652173913043
Food & Drink : 0.8074534161490683
Reference : 0.5590062111801243
Business : 0.5279503105590062
Book : 0.43478260869565216
Navigation : 0.18633540372670807
Medical : 0.18633540372670807
Catalogs : 0.12422360248447205


In [42]:
display_table(android_final,9)

Tools : 8.450863138892023
Entertainment : 6.070179397495204
Education : 5.348076272142616
Business : 4.592124562789123
Productivity : 3.8925871601038025
Lifestyle : 3.8925871601038025
Finance : 3.7007785174320205
Medical : 3.5315355974275078
Sports : 3.463838429425702
Personalization : 3.317161232088458
Communication : 3.2381812027530184
Action : 3.102786866749408
Health & Fitness : 3.0802211440821394
Photography : 2.944826808078529
News & Magazines : 2.798149610741284
Social : 2.6627552747376737
Travel & Local : 2.324269434728647
Shopping : 2.245289405393208
Books & Reference : 2.1437436533904997
Simulation : 2.042197901387792
Dating : 1.8616721200496444
Arcade : 1.8503892587160102
Video Players & Editors : 1.771409229380571
Casual : 1.7601263680469368
Maps & Navigation : 1.399074805370642
Food & Drink : 1.241114746699763
Puzzle : 1.128286133363421
Racing : 0.9928917973598104
Role Playing : 0.9364774906916393
Libraries & Demo : 0.9364774906916393
Auto & Vehicles : 0.9251946293580051
S

##### Most common genres in iOS

The most common genre is Tools. The runner-up is Entertainment. General impression is that most of the apps are  designed for  for entertainment (games, photo and video, social networking, sports, music).Based on this can draw rough conclusion on popularity of the apps on that genre. Can draw conclusion by based on number of apps would mean that there are a lot of users for that genre. 

##### Most common genres in Android Play store

The most common genre is Tools.The runner-up is Entertainment.General impression is that most of the apps are designed for for practical purposes (education, shopping, utilities, productivity, lifestyle). For Android, the distribution of downloads is more evenly distributed. 

# Most Popular Apps by Genre

Now, we'd like to get an idea about the kind of apps with the most users. One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

We'll start with calculating the average number of user ratings per app genre on the App Store. To do that, we'll need to:
- Isolate the apps of each genre.
- Sum up the user ratings for the apps of that genre.
- Divide the sum by the number of apps belonging to that genre (not by the total number of apps).

##### Most Popular Apps in iOS

apple_freq = freq_table(apple_final,-5)

for genre in apple_freq:
    total = 0
    len_genre = 0

    for app in apple_final:
        genre_app = app[-5]
        if genre_app == genre:
            app_ratings = float(app[6])
            total += app_ratings
            len_genre += 1

    average = total / len_genre
    print(genre, ':', average)

Based on the above, it appears that Navigation is the most popular Genre by downloads for the App Store. 

In [43]:
for app in apple_final:
    if app[-5] == 'Navigation':
        print(app[2], ':', app[6]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Geocaching® : 12811
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5
CoPilot GPS – Car Navigation & Offline Maps : 3582
Google Maps - Navigation & Transit : 154911


From the above, Waze and Google Maps skewed the distribution with 90% of the downloads of Navigation apps. 

In [44]:
for app in apple_final:
    if app[-5] == 'Reference':
        print(app[2], ':', app[6]) # print name and number of ratings

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
Merriam-Webster Dictionary : 16849
Google Translate : 26786
Night Sky : 12122
WWDC : 762
Jishokun-Japanese English Dictionary & Translator : 0
教えて!goo : 0
VPN Express : 14
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Real Bike Traffic Rider Virtual Reality Glasses : 8


For Reference, religious texts and dictionaries are the highest number of downloads. However, there is potential if we take a popular book and turn it into an app. Let's see if this pattern repeats in Google Play. 

##### Most Popular Apps by Genre on Google Play

In [45]:
android_freq = freq_table(android_final,1)
android_freq

for category in android_freq:
    total = 0
    len_genre = 0

    for app in android_final:
        category_app = app[1]
        if category_app == category:
            app_install = app[5]
            app_install = app_install.replace('+','')
            app_install = app_install.replace(',','')
            app_install = float(app_install)
            total += app_install
            len_genre += 1

    average = total / len_genre

    print(category, ':', average)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3697848.1731343283
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

Based on the above, it appears that Communication is the most popular Category by downloads for the Play Store. 

In [46]:
for app in android_final:
    if app[1] == 'COMMUNICATION':
        print(app[0], ':', app[5]) # print name and number of installs

WhatsApp Messenger : 1,000,000,000+
Messenger for SMS : 10,000,000+
My Tele2 : 5,000,000+
imo beta free calls and text : 100,000,000+
Contacts : 50,000,000+
Call Free – Free Call : 5,000,000+
Web Browser & Explorer : 5,000,000+
Browser 4G : 10,000,000+
MegaFon Dashboard : 10,000,000+
ZenUI Dialer & Contacts : 10,000,000+
Cricket Visual Voicemail : 10,000,000+
TracFone My Account : 1,000,000+
Xperia Link™ : 10,000,000+
TouchPal Keyboard - Fun Emoji & Android Keyboard : 10,000,000+
Skype Lite - Free Video Call & Chat : 5,000,000+
My magenta : 1,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Seznam.cz : 1,000,000+
Antillean Gold Telegram (original version) : 100,000+
AT&T Visual Voicemail : 10,000,000+
GMX Mail : 10,000,000+
Omlet Chat : 10,000,000+
My Vodacom SA : 5,000,000+
Microsoft Edge : 5,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Calls & Text by Mo+ : 5,000,000+
free 

We can tell here that for Android, messaging apps are most popular in downloads. If we are to look at the number of apps with more than 100M downloads. 

In [47]:
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we remove the outliers of above 100M, let's see how many average installs there are for Communication genre.

In [48]:
under_100_m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
print("Avg Communication:",sum(under_100_m) / len(under_100_m))

Avg Communication: 3603485.3884615386


It looks like there are a few Giants dominating for Communication genre which makes it hard to compete. Let's look at  another genre, which is video players. 

In [49]:
for app in android_final:
    if app[1] == 'VIDEO_PLAYERS':
        print(app[0], ':', app[5]) # print name and number of installs

YouTube : 1,000,000,000+
All Video Downloader 2018 : 1,000,000+
Video Downloader : 10,000,000+
HD Video Player : 1,000,000+
Iqiyi (for tablet) : 1,000,000+
Video Player All Format : 10,000,000+
Motorola Gallery : 100,000,000+
Free TV series : 100,000+
Video Player All Format for Android : 500,000+
VLC for Android : 100,000,000+
Code : 10,000,000+
Vote for : 50,000,000+
XX HD Video downloader-Free Video Downloader : 1,000,000+
OBJECTIVE : 1,000,000+
Music - Mp3 Player : 10,000,000+
HD Movie Video Player : 1,000,000+
YouCut - Video Editor & Video Maker, No Watermark : 5,000,000+
Video Editor,Crop Video,Movie Video,Music,Effects : 1,000,000+
YouTube Studio : 10,000,000+
video player for android : 10,000,000+
Vigo Video : 50,000,000+
Google Play Movies & TV : 1,000,000,000+
HTC Service － DLNA : 10,000,000+
VPlayer : 1,000,000+
MiniMovie - Free Video and Slideshow Editor : 50,000,000+
Samsung Video Library : 50,000,000+
OnePlus Gallery : 1,000,000+
LIKE – Magic Video Maker & Community : 50,

In [50]:
for app in android_final:
    if app[1] == 'VIDEO_PLAYERS' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

YouTube : 1,000,000,000+
Motorola Gallery : 100,000,000+
VLC for Android : 100,000,000+
Google Play Movies & TV : 1,000,000,000+
MX Player : 500,000,000+
Dubsmash : 100,000,000+
VivaVideo - Video Editor & Photo Movie : 100,000,000+
VideoShow-Video Editor, Video Maker, Beauty Camera : 100,000,000+
Motorola FM Radio : 100,000,000+


And after if we remove the outliers. 

In [51]:
under_100_m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'VIDEO_PLAYERS') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
print("Avg Video Players:",sum(under_100_m) / len(under_100_m))

Avg Video Players: 5544878.133333334


Repeating the same observation, the market for video players is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.). Books and reference looks pretty popular. Let's see if there is a niche we can fit into. 

In [52]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5]) # print name and number of installs

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

In [53]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


It looks like there are not that many outliers, so there can be potential here. Let's look at apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [54]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

There are quite a number of religious texts here. However, there is still potential by taking a popular book or idea and turn it into an app, e.g. My Little Pony AR Guide. 

# Summary

From the above, we analysed the apps in App Store, and Google Play and conclude that books genre can be worth exploring as it is not too saturated with a few heavyweights, and we can possibly turn a popular idea to become an app. 